fv3gfs/physics/stencils/physics.py

```diff
@@ -56,6 +59,106 @@ def atmos_phys_driver_statein(
     with computation(PARALLEL), interval(0, 1):
         prsi = ptop
 
+    with computation(FORWARD), interval(1, None):
+        prsi = prsi[0, 0, -1] + delp[0, 0, -1]
+
+    with computation(PARALLEL), interval(0, -1):
+        prsik = log(prsi)
+        qvapor = qvapor / delp
+        qliquid = qliquid / delp
+        qrain = qrain / delp
+        qice = qice / delp
+        qsnow = qsnow / delp
+        qgraupel = qgraupel / delp
+        qo3mr = qo3mr / delp
+        qsgs_tke = qsgs_tke / delp
+
+    with computation(PARALLEL), interval(-1, None):
+        prsik = log(prsi)
+
+    with computation(PARALLEL), interval(0, 1):
+        prsik = log(ptop)
+
+    with computation(PARALLEL), interval(0, -1):
+        qmin = 1.0e-10  # set it here since externals cannot be 2D
+        qgrs_rad = max(qmin, qvapor)
+        rTv = rdgas * pt * (1.0 + con_fvirt * qgrs_rad)
+        dm = delp[0, 0, 0]
+        delp = dm * rTv / (phii[0, 0, 0] - phii[0, 0, 1])
+        delp = min(delp, prsi[0, 0, 1] - 0.01 * dm)
+        delp = max(delp, prsi + 0.01 * dm)
+
+    with computation(PARALLEL), interval(-1, None):
+        prsik = exp(KAPPA * prsik) * pk0inv
+
+    with computation(PARALLEL), interval(0, 1):
+        prsik = pktop
+
+
+def prepare_microphysics(
+    dz: FloatField,
+    phii: FloatField,
+    wmp: FloatField,
+    omga: FloatField,
+    qvapor: FloatField,
+    pt: FloatField,
+    delp: FloatField,
+):
+    with computation(BACKWARD), interval(...):
+        dz = (phii[0, 0, 1] - phii[0, 0, 0]) * rgrav
+        wmp = -omga * (1.0 + con_fvirt * qvapor) * pt / delp * (rdgas * rgrav)
+
+
+@gtscript.function
+def forward_euler(q_t0, q_dt, dt):
+    return q_t0 + q_dt * dt
+
+
+def update_physics_state_with_tendencies(
+    qvapor: FloatField,
+    qliquid: FloatField,
+    qrain: FloatField,
+    qice: FloatField,
+    qsnow: FloatField,
+    qgraupel: FloatField,
+    qcld: FloatField,
+    pt: FloatField,
+    ua: FloatField,
+    va: FloatField,
+    qv_dt: FloatField,
+    ql_dt: FloatField,
+    qr_dt: FloatField,
+    qi_dt: FloatField,
+    qs_dt: FloatField,
+    qg_dt: FloatField,
+    qa_dt: FloatField,
+    pt_dt: FloatField,
+    udt: FloatField,
+    vdt: FloatField,
+    qvapor_t1: FloatField,
+    qliquid_t1: FloatField,
+    qrain_t1: FloatField,
+    qice_t1: FloatField,
+    qsnow_t1: FloatField,
+    qgraupel_t1: FloatField,
+    qcld_t1: FloatField,
+    pt_t1: FloatField,
+    ua_t1: FloatField,
+    va_t1: FloatField,
+    dt: Float,
+):
+    with computation(PARALLEL), interval(...):
+        qvapor_t1 = forward_euler(qvapor, qv_dt, dt)
```

Comment:
It would be great if we could do gt4py for loops!

Reply:
Yes. In fact, I think we can now if we put the tracers into a higher dimensional field. Except we need to move physics to gtc backend, hopefully soon...